### PR TITLE
add task_type_id to BERT to support ERNIE-2.0 and ERNIE-3.0 models

### PR DESCRIPTION
**Pull request by nghuyong**
# What does this PR do?

[ERNIE2.0](https://arxiv.org/abs/1907.12412) citation and [ERNIE3.0](https://arxiv.org/abs/2107.02137) citation  are a series of powerful models based on BERT, especially in Chinese tasks. These models introduce `task_type_embeddings` in the embedding layer, so this PR is to support this feature.

the config of ERNIE2.0 / ERNIE3.0 models have the following two params:
```json
...
"task_type_vocab_size": 3,
"use_task_id": true
...
```
the released ERNIE2.0 / ERNIE3.0 models have the weight of `bert.embeddings.task_type_embeddings.weight`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.  https://github.com/huggingface/transformers/issues?q=is%3Aissue+ernie
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?
I write a [script](https://github.com/nghuyong/ERNIE-Pytorch/blob/master/convert_ernie3.0.py) to convert the official released [ERNIE3.0 model (paddlepaddle version)](https://github.com/PaddlePaddle/PaddleNLP/tree/develop/model_zoo/ernie-3.0). And I have checked that the model results before and after transformation are consistent (with `task_type_embedding` added)

```Python
import paddle
import torch
from paddlenlp.transformers import ErnieForMaskedLM, ErnieTokenizer
# this PR version
from transformers import BertTokenizer, BertForMaskedLM

tokenizer = BertTokenizer.from_pretrained('nghuyong/ernie-3.0-base-zh')
model = BertForMaskedLM.from_pretrained('nghuyong/ernie-3.0-base-zh')
input_ids = torch.tensor([tokenizer.encode(text="[MASK][MASK][MASK]是中国神魔小说的经典之作，与《三国演义》《水浒传》《红楼梦》并称为中国古典四大名著。",
                                           add_special_tokens=True)])
model.eval()
with torch.no_grad():
    predictions = model(input_ids)[0][0]
predicted_index = [torch.argmax(predictions[i]).item() for i in range(predictions.shape[0])]
predicted_token = [tokenizer._convert_id_to_token(predicted_index[i]) for i in
                   range(1, (predictions.shape[0] - 1))]
print('huggingface result')
print('predict result:\t', predicted_token)
print('[CLS] logit:\t', predictions[0].numpy())
tokenizer = ErnieTokenizer.from_pretrained("ernie-3.0-base-zh")
model = ErnieForMaskedLM.from_pretrained("ernie-3.0-base-zh")
inputs = tokenizer("[MASK][MASK][MASK]是中国神魔小说的经典之作，与《三国演义》《水浒传》《红楼梦》并称为中国古典四大名著。")
inputs = {k: paddle.to_tensor([v]) for (k, v) in inputs.items()}
model.eval()
with paddle.no_grad():
    predictions = model(**inputs)[0]
predicted_index = [paddle.argmax(predictions[i]).item() for i in range(predictions.shape[0])]
predicted_token = [tokenizer._convert_id_to_token(predicted_index[i]) for i in
                   range(1, (predictions.shape[0] - 1))]
print('paddle result')
print('predict result:\t', predicted_token)
print('[CLS] logit:\t', predictions[0].numpy())

“”“
huggingface result
predict result:	 ['西', '游', '记', '是', '中', '国', '神', '魔', '小', '说', '的', '经', '典', '之', '作', '，', '与', '《', '三', '国', '演', '义', '》', '《', '水', '浒', '传', '》', '《', '红', '楼', '梦', '》', '并', '称', '为', '中', '国', '古', '典', '四', '大', '名', '著', '。']
[CLS] logit:	 [-20.574057  -29.192085  -15.638802  ...  -1.9127564  -1.4329851  -1.8172828]

paddle result
predict result:	 ['西', '游', '记', '是', '中', '国', '神', '魔', '小', '说', '的', '经', '典', '之', '作', '，', '与', '《', '三', '国', '演', '义', '》', '《', '水', '浒', '传', '》', '《', '红', '楼', '梦', '》', '并', '称', '为', '中', '国', '古', '典', '四', '大', '名', '著', '。']
[CLS] logit:	 [-20.573637  -29.193172  -15.639115  ...  -1.9127647  -1.4330447  -1.816982 ]
”“”
```

## Who can review?

@LysandreJik

### CODE REVIEW

**Comment by LysandreJik** [comment]
Hey @nghuyong, thanks for your PR! 

In this situation, we'd rather have a new model class "Ernie" rather than modifying the "Bert" model class. This will result in a larger PR, but it should be very little additional work for you.

I encourage you to follow the following guide: https://github.com/huggingface/transformers/tree/main/templates/adding_a_new_model#add-new-model-like-command

It seems like it would be as simple as using the scfript to add a new model like BERT, but with the ERNIE name; then applying the changes above.

@ydshieh, would you be down to help @nghuyong if they run into any problem? 

Thanks a lot!

**Comment by nghuyong** [comment]
Thanks for your advice. I will try to do this work.

**Comment by ydshieh** [comment]
Agree with @LysandreJik to have a new model file for this. And it should be fairly straightforward.

Glad to know you already have the checkpoints available! Let me me if you need any help, @nghuyong for the PR.
Looking forward to review it! Thanks, @nghuyong !

**Comment by nghuyong** [comment]
@ydshieh @LysandreJik  I have updated this PR and add Ernie model.
Please help to review it, if you have any questions, you can AT me, thanks!

**Comment by LysandreJik** [comment]
Wonderful, thank you @nghuyong!

**Comment by ydshieh** [comment]
Hi @nghuyong, Great job! I will review the PR. Currently, the failing tests are caused by 

```bash
E   AttributeError: module transformers.models.ernie has no attribute BasicTokenizer
```
which is from your change in
```
src/transformers/__init__.py
```

I will discuss my colleagues to see how should we do for the tokenizers.

File: src/transformers/__init__.py
Comment (ydshieh) [inline src/transformers/__init__.py:148]: Let's remove this to see if the test results get better :-)

File: src/transformers/__init__.py
Comment (ydshieh) [inline src/transformers/__init__.py:151]: Same here

File: src/transformers/__init__.py
Comment (ydshieh) [inline src/transformers/__init__.py:2978]: same here

File: src/transformers/__init__.py
Comment (ydshieh) [inline src/transformers/__init__.py:2981]: same here

File: src/transformers/models/ernie/modeling_ernie.py
Comment (ydshieh) [inline src/transformers/models/ernie/modeling_ernie.py:83]: I think we have (at least some) checkpoints for ERNIE 2/3, is this correct? Could we use them here maybe?

Comment (nghuyong) [reply src/transformers/models/ernie/modeling_ernie.py:83]: Yes, now we have the checkpoints for ERNIE1.0 ~3.0, all checkpoint can be seen [here](https://github.com/nghuyong/ERNIE-Pytorch#supported-models)

But these checkpoints are all BertForMaskedLM or BertModel model.
So, I comment _CHECKPOINT_FOR_SEQUENCE_CLASSIFICATION

**Comment by nghuyong** [comment]
@ydshieh  So,  is there anything that needs to be updated now?

**Comment by ydshieh** [comment]
Hi @nghuyong Thanks a lot :-). I will take a full review this week.

The current failing tests (most of them) could be fixed by updating your branch. You can do it as
```python
git checkout main
git pull upstream main
git checkout [YOUR-WORKING-BRANCH]
git rebase main
git push --force-with-lease
```

Before doing so, it would be a good idea to keep a backup of current branch in case of the commit history being messed up.
```python
git checkout add_task_type_id
git checkout -b add_task_type_id_backup
```

**Comment by nghuyong** [comment]
Thanks, @ydshieh my branch has been synced with master now

**Comment by HUSTHY** [comment]
@nghuyong 
你好，你的ernie的分支是已经合并到mater了吗——意思是安装最新的transformers库，就可以直接使用ernie3和你提供的ernie3的中文权重吧；如不需要通过之前重装transformers的方式：
pip install git+https://github.com/nghuyong/transformers@add_task_type_id

**Comment by nghuyong** [comment]
@HUSTHY, still not, and  `add_task_type_id` has been changed, so, you cannot load ernie3 now.

**Comment by HUSTHY** [comment]
已收到！谢谢！                          ——黄洋

**Comment by HUSTHY** [comment]
@nghuyong  好吧 我还以为能用了呢  那你实现的这个分支的代码是没有问题的吧。。。 我直接把代码放在本地应该OK的

**Review by ydshieh — COMMENTED** [review]
Thank you a lot @nghuyong for adding this model. It is in great shape already, and the review is quite smooth 🤗 .

I left some comments to be addressed.

Once everything is ready, we have a few steps to take before merge. For example,
```
make fix-copies
```
is required. But we can do this at the final commit(s).

**Comment by ydshieh** [comment]
@nghuyong When you add argument `task_type_ids` to the other model classes, you will also need to remove some `# copied from`, for example this part

```
# Copied from transformers.models.bert.modeling_bert.BertModel with BERT->ERNIE,Bert->Ernie
class ErnieModel(ErniePreTrainedModel):
```

You can instead do the following (and same for other methods), although it's more tedious
```
    # Copied from transformers.models.bert.modeling_bert.BertModel.__init__ with BERT->ERNIE,Bert->Ernie
    def __init__(self, config, add_pooling_layer=True):
```

File: src/transformers/models/auto/modeling_auto.py
Comment (nghuyong) [reply src/transformers/models/auto/modeling_auto.py:698]: don't need to change

**Comment by nghuyong** [comment]
@ydshieh Thanks for your review!! I have updated following your advice.

File: src/transformers/models/ernie/modeling_ernie.py
Comment (ydshieh) [inline src/transformers/models/ernie/modeling_ernie.py:656]: This has to be removed in order for `make fix-copies` not to complain.

This is because BERT has a line

```
    load_tf_weights = load_tf_weights_in_bert
```
but not in ERNIE.

File: src/transformers/models/ernie/modeling_ernie.py
Comment (ydshieh) [inline src/transformers/models/ernie/modeling_ernie.py:503]:
```suggestion
            if self.gradient_checkpointing and self.training:
            
                if use_cache:
```

Comment (ydshieh) [reply src/transformers/models/ernie/modeling_ernie.py:503]: Need an extra empty line. In fact, you can run `make fix-copies` to detect the necessary changes. But you can commit directly here.

**Review by ydshieh — CHANGES_REQUESTED** [review]

**Comment by ydshieh** [comment]
Thank you, @nghuyong. Here are a few remaining step from my side

- First, address the above 2 review comments.

- Then
```bash
make style
```
it will fix
```
transformers\src\transformers\__init__.py
transformers\src\transformers\models\ernie\__init__.py
```

- Finally
```bash
pip install -U huggingface_hub
make fix-copies
```
to fix some README files.

Then we can have the core maintainers to have a final look 🚀

**Comment by ydshieh** [comment]
For the style, could you try `pip install -U hf-doc-builder` before `make style`

**Comment by ydshieh** [comment]
If there are still issues, I can make the change and push to this PR directly, if you are OK with this

**Comment by nghuyong** [comment]
Hi @ydshieh , I think my development env may be wrong.
First,  I create a new conda env (`conda create -n hf python=3.9`) and run `$ pip install -e ".[dev]"`
get this error
```
ERROR: Could not find a version that satisfies the requirement tensorflow<2.10,>=2.3; extra == "dev" (from transformers[dev]) (from versions: none)
ERROR: No matching distribution found for tensorflow<2.10,>=2.3; extra == "dev"
```
Then, in this new env, I run:
```
pip install -U hf-doc-builder
make style
```
and get this error
```
black --preview examples tests src utils
make: black: No such file or directory
make: *** [style] Error 1
```

ps, my chip is Apple M1 Pro

**Comment by ydshieh** [comment]
Glad it works for you. It's probably not working well TF + Mac (not very sure, I rarely use this combination).
I think you haven't run this yet
```
pip install -U huggingface_hub
make fix-copies
```
right? It's necessary to pass the `check_repository_consistency` CI.

**Comment by nghuyong** [comment]
Yes, Is it time to run `make fix-copies` ? @ydshieh

**Comment by ydshieh** [comment]
Thanks a lot. I will take care the remaining issue in the CI tomorrow. Somehow strange here.

**Comment by nghuyong** [comment]
@ydshieh  Because I use `ERNIE` not `Ernie` in Readme. so the ci get failed

**Comment by ydshieh** [comment]
You can ignore the test failure in `ci/circleci: run_tests_hub`.

**Comment by nghuyong** [comment]
@ydshieh  OK,   `run_tests_hub` could be ignored, may be the ci has no problems now.

**Comment by ydshieh** [comment]
Hi @nghuyong I push a commit which adds several `# copied from ...`

**Comment by ydshieh** [comment]
@sgugger This is model is just BERT, but with a new argument `task_type_ids` which is used to create a new embedding to be summed. Time for you to have a final review 🙏

File: src/transformers/__init__.py
Comment (sgugger) [inline src/transformers/__init__.py:905]:
```suggestion
```
This one should not be imported here. It will be accessible via the submodule.

File: src/transformers/models/ernie/modeling_ernie.py
Comment (sgugger) [inline src/transformers/models/ernie/modeling_ernie.py:3]: Let's adapt the copyright here.
```suggestion
# Copyright 2022 The HuggingFace Inc. team.
```

File: docs/source/en/model_doc/ernie.mdx
Comment (sgugger) [inline docs/source/en/model_doc/ernie.mdx:21]: It would be nice to have an example of use here.

File: src/transformers/models/ernie/modeling_ernie.py
Comment (sgugger) [inline src/transformers/models/ernie/modeling_ernie.py:766]: I don't fully understand what this means, and since this is the new argument from BERT, it would be great to have more information here!

File: src/transformers/models/ernie/modeling_ernie.py
Comment (sgugger) [inline src/transformers/models/ernie/modeling_ernie.py:1095]: This name was used before, but we now prefer the clearer `ErnieForCausalLM`.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! I left a couple of comments, mainly around documentation.
Also please rename everywhere `ErnieLMHeadModel` to `ErnieForCausalLM` (we can't do the rename for BERT for backward compatibility reasons, but for new models we prefer this terminology).

File: src/transformers/__init__.py
Comment (sgugger) [reply src/transformers/__init__.py:905]: Yes BertLayer shouldn't have been exposed as well, but we won't remove it now for BC reasons.

**Comment by nghuyong** [comment]
@ydshieh @sgugger  I have updated the code, but don't pass the ci.  need some help :)

**Review by sgugger — APPROVED** [review]

**Review by ydshieh — APPROVED** [review]

**Comment by ydshieh** [comment]
Thank! Pushed a [final commit](https://github.com/huggingface/transformers/pull/18686/commits/685728232335792f1746f2a5772e60545a545810) to remove the remaining `ErnieLMHeadModel` in a log message.

@sgugger I approved :-)

**Comment by ydshieh** [comment]
failed test is irrelevant to this PR.

**Pull request merged**
